routing/http/server/server.go

```diff
@@ -132,7 +147,145 @@ type server struct {
 	streamingRecordsLimit int
 }
 
+func (s *server) detectResponseType(r *http.Request) (string, error) {
+	var (
+		supportsNDJSON bool
+		supportsJSON   bool
+
+		acceptHeaders = r.Header.Values("Accept")
+	)
+
+	if len(acceptHeaders) == 0 {
+		return mediaTypeJSON, nil
+	}
+
+	for _, acceptHeader := range acceptHeaders {
+		for _, accept := range strings.Split(acceptHeader, ",") {
+			mediaType, _, err := mime.ParseMediaType(accept)
+			if err != nil {
+				return "", fmt.Errorf("unable to parse Accept header: %w", err)
+			}
+
+			switch mediaType {
+			case mediaTypeJSON, mediaTypeWildcard:
+				supportsJSON = true
+			case mediaTypeNDJSON:
+				supportsNDJSON = true
+			}
+		}
+	}
+
+	if supportsNDJSON && !s.disableNDJSON {
+		return mediaTypeNDJSON, nil
+	} else if supportsJSON {
+		return mediaTypeJSON, nil
+	} else {
+		return "", errors.New("no supported content types")
+	}
+}
+
+func (s *server) findProviders(w http.ResponseWriter, httpReq *http.Request) {
+	vars := mux.Vars(httpReq)
+	cidStr := vars["cid"]
+	cid, err := cid.Decode(cidStr)
+	if err != nil {
+		writeErr(w, "FindProviders", http.StatusBadRequest, fmt.Errorf("unable to parse CID: %w", err))
+		return
+	}
+
+	mediaType, err := s.detectResponseType(httpReq)
+	if err != nil {
+		writeErr(w, "FindProviders", http.StatusBadRequest, err)
+		return
+	}
+
+	var (
+		handlerFunc  func(w http.ResponseWriter, provIter iter.ResultIter[types.Record])
+		recordsLimit int
+	)
+
+	if mediaType == mediaTypeNDJSON {
+		handlerFunc = s.findProvidersNDJSON
+		recordsLimit = s.streamingRecordsLimit
+	} else {
+		handlerFunc = s.findProvidersJSON
+		recordsLimit = s.recordsLimit
+	}
+
+	provIter, err := s.svc.FindProviders(httpReq.Context(), cid, recordsLimit)
+	if err != nil {
+		writeErr(w, "FindProviders", http.StatusInternalServerError, fmt.Errorf("delegate error: %w", err))
+		return
+	}
+
+	handlerFunc(w, provIter)
+}
+
+func (s *server) findProvidersJSON(w http.ResponseWriter, provIter iter.ResultIter[types.Record]) {
+	defer provIter.Close()
+
+	providers, err := iter.ReadAllResults(provIter)
+	if err != nil {
+		writeErr(w, "FindProviders", http.StatusInternalServerError, fmt.Errorf("delegate error: %w", err))
+		return
+	}
+
+	writeJSONResult(w, "FindProviders", jsontypes.ProvidersResponse{
+		Providers: providers,
+	})
+}
+
+func (s *server) findProvidersNDJSON(w http.ResponseWriter, provIter iter.ResultIter[types.Record]) {
+	writeResultsIterNDJSON(w, provIter)
+}
+
+func (s *server) findPeers(w http.ResponseWriter, r *http.Request) {
+	pidStr := mux.Vars(r)["peer-id"]
+
+	// pidStr must be in CIDv1 format. Therefore, use [cid.Decode]. We can't use
+	// [peer.Decode] because that would allow other formats to pass through.
+	cid, err := cid.Decode(pidStr)
+	if err != nil {
+		writeErr(w, "FindPeers", http.StatusBadRequest, fmt.Errorf("unable to parse peer ID: %w", err))
+		return
+	}
```

Comment:
:warning: UX hiccup:  this error will be really, really annoying in Kubo, where commands like `ipfs swarm peers` return non-CID PeerIDs like `12D3KooWHj48PbuAPUf9LeovBJZEXn2A4rseW5z9wbTBwvM82nAq`.
User will copy it, and then get this error without any 


```console
$ curl  http://127.0.0.1:8280/routing/v1/peers/12D3KooWHj48PbuAPUf9LeovBJZEXn2A4rseW5z9wbTBwvM82nAq
unable to parse peer ID: invalid cid: selected encoding not supported
```

@hacdias I agree we should error here, but could you improve error mesaage?

Do what we already do at https://cid.ipfs.tech/#12D3KooWHj48PbuAPUf9LeovBJZEXn2A4rseW5z9wbTBwvM82nAq and return meaningful error that has CIDv1 representation and asks user to try again with it (actionable, allows user to self-service and fix without asking on forums, googling).